app/heron/factories/tube_rack.rb

```diff
@@ -9,28 +9,40 @@ class TubeRack
       include Concerns::HeronStudy
       include Concerns::CoordinatesSupport
       include Concerns::ForeignBarcodes
+      include Concerns::RecipientsCoordinates
+      include Concerns::Recipients
+      include Concerns::Contents
 
-      attr_accessor :sample_tubes, :tube_rack, :size
+      attr_accessor :sample_tubes, :tube_rack
 
-      validates_presence_of :tubes, :size, :purpose
-      validate :check_tubes
+      validates_presence_of :size, :purpose, :recipients
 
-      def tubes=(attributes)
-        attributes.each do |tube|
-          tubes[tube[:coordinate]] = ::Heron::Factories::Tube.new(tube.except(:coordinate).merge(study: heron_study))
-        end
+      def initialize(params)
+        @params = params
       end
 
-      def racked_tubes
-        @racked_tubes ||= []
+      def recipients_key
+        :tubes
       end
 
-      def purpose
-        @purpose ||= ::TubeRack::Purpose.where(target_type: 'TubeRack', size: size).first
+      def recipient_factory
+        ::Heron::Factories::Tube
+      end
+
+      def content_factory
```

Comment:
Not sure content_factory has to be abstracted - will it ever be anything other than Sample?

Reply:
Not sure. I've added it to be consistent with recipient_factory for recipients. My expectations is that none of them are going to change